server/src/main/java/org/elasticsearch/snapshots/package-info.java

```diff
@@ -0,0 +1,112 @@
+ * <ol>
+ * <li>An invocation of {@link org.elasticsearch.snapshots.SnapshotsService#createSnapshot} enqueues a cluster state update to create
+ * a {@link org.elasticsearch.cluster.SnapshotsInProgress.Entry} in the cluster-state's {@code SnapshotsInProgress}.
+ * This initial snapshot entry has its state set to {@code INIT} and an empty map set for the state of the individual shard's snapshots.
+ * </li>
+ *
+ * <li>After the snapshot's entry with state {@code INIT} is in the cluster-state,
+ * {@link org.elasticsearch.snapshots.SnapshotsService} determines the primary shards' assignments for all indices that are being
+ * snapshotted and updates the existing {@code SnapshotsInProgress.Entry} with state {@code STARTED} and adds the map of
+ * {@link org.elasticsearch.index.shard.ShardId} to {@link org.elasticsearch.cluster.SnapshotsInProgress.ShardSnapshotStatus} that tracks
+ * the assignment of which node is to snapshot which shard. All shard snapshots are executed on the shard's primary node.
+ * Thus all shards for which the primary node was found to have a healthy copy of the shard are marked as being in state {@code INIT} in
+ * this map.
+ * If the primary for a shard is unassigned, it is marked as {@code MISSING} in this
+ * map. In case the primary is initializing at this point, it is marked as in state
+ * {@code WAITING}.</li>
+ *
+ * <li>The new {@code SnapshotsInProgress.Entry} is then observed by
+ * {@link org.elasticsearch.snapshots.SnapshotShardsService#clusterChanged} on all nodes and since the entry is in state
+ * {@code STARTED} the {@code SnapshotShardsService} will check if any local primary shards are to be snapshotted
+ * (signaled by the shard's snapshot state being {@code INIT}).
+ * For those local primary shards found in state {@code INIT}) the snapshot process
+ * of writing the shard's data files to the snapshot's {@link org.elasticsearch.repositories.Repository} is executed.
+ * Once the snapshot execution finishes for a shard an {@code UpdateIndexShardSnapshotStatusRequest} is sent to the master node signaling
+ * either status {@code SUCCESS} or {@code FAILED}.
+ * The master node will then update a shard's state in the snapshots {@code SnapshotsInProgress.Entry} whenever
+ * it receives such a {@code UpdateIndexShardSnapshotStatusRequest}.</li>
+ *
+ * <li>If as a result of the received status update requests, all shards in the cluster-state are in a completed state, i.e are marked as
+ * either {@code SUCCESS}, {@code FAILED} or {@code MISSING}, the {@code SnapshotShardsService}
+ * will update the state of the {@code Entry} itself and mark it as {@code SUCCESS}.
+ * At the same time {@link org.elasticsearch.snapshots.SnapshotsService#endSnapshot} is executed, writing the metadata necessary to
+ * finalize the snapshot in the repository to the repository.</li>
+ *
+ * <li>After writing the final metadata to the repository, a cluster-state update to remove the snapshot from the repository is submitted
```

Comment:
I think this should be:

```suggestion
 * <li>After writing the final metadata to the repository, a cluster-state update to remove the snapshot from the cluster state is submitted
```